### PR TITLE
2. fix(state): index spending transaction IDs for each address

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

When I compare `zebrad` and `zcashd` RPC responses for `getaddresstxids`, using this command:
```sh
zebra-utils/zcash-rpc-diff 28232 'getaddresstxids { "addresses": ["t3dvVE3SQEi7kqNzwrfNePxZ1d4hUyztBA1"], "start": 1, "end": 1661134 }'
```

All the spending transaction IDs are missing.

Depends-On: #4385

### Specifications

It's not explicitly listed in the docs, but `zcashd` includes spending transaction IDs:
https://zcash.github.io/rpc/getaddresstxids.html

## Solution

- Correctly populate spending address transaction ID indexes
- Increment the database format to rebuild address tx ID indexes

Related changes:
- Split writing transaction indexes into transparent and shielded
- Split writing transparent indexes into created and spent
- Update non-finalized docs to prevent similar bugs 

@gustavovalverde @dconnolly this PR might cause DevOps bugs or slow builds, because it increments the state version.

## Review

@conradoplg and I wrote this code.

I've marked this PR as a high priority, because it changes the state version.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

It's really tricky to test this code, because transparent outputs can't be spent in the first 100 blocks, due to the coinbase maturity consensus rule. I'm testing it locally by doing a full sync, then re-running the manual comparison with `zcashd`.

### CODE REVIEW

File: zebra-state/src/service/finalized_state/zebra_db/transparent.rs [hunk 1]
```diff
@@ -438,31 +515,71 @@ impl DiskWriteBatch {
...
                 );
                 self.zs_delete(&utxo_loc_by_transparent_addr_loc, address_spent_output);
-
-                // Create a link from the AddressLocation to the spent TransactionLocation in the database.
-                // Unlike the OutputLocation link, this will never be deleted.
-                let address_transaction = AddressTransaction::new(
-                    sending_address_location,
-                    spent_output_location.transaction_location(),
-                );
-                self.zs_insert(&tx_loc_by_transparent_addr_loc, address_transaction, ());
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/transparent.rs:458]: This is the location of the transaction that created the UTXO, but we need to index the spending transaction.

File: zebra-state/src/service/finalized_state/zebra_db/transparent.rs [hunk 2]
```diff
@@ -438,31 +515,71 @@ impl DiskWriteBatch {
...
+                    .address_location();
+
+                // Create a link from the AddressLocation to the spent TransactionLocation in the database.
+                // Unlike the OutputLocation link, this will never be deleted.
+                //
+                // The value is the location of this transaction,
+                // not the transaction the spent output is from.
+                let address_transaction =
+                    AddressTransaction::new(sending_address_location, spending_tx_location);
+                self.zs_insert(&tx_loc_by_transparent_addr_loc, address_transaction, ());
```
Comment (teor2345) [inline zebra-state/src/service/finalized_state/zebra_db/transparent.rs:580]: To fix this issue, we pass the location of the spending transaction as an argument, and use it in the index.

**Comment by teor2345** [comment]
I tested the comparison locally, and I got:
> Response diff between zcashd and zebrad:
> RPC responses were identical

**Comment by teor2345** [comment]
CI is not rebuilding the cached state in this PR, even though the state version has changed.

So these tests failed:
- fully-synced-rpc 
- lightwalletd-transactions

This was meant to be fixed as part of #4153, but that's a big ticket. Maybe I can do a smaller fix in this PR.

Or we could approve this PR and merge it to `main`. Then the cached state will get rebuilt, and the tests should work.

**Review by conradoplg — DISMISSED** [review]
Took me a while to grok this (I always get confused by inputs/outputs) but I get it now, thanks!

I was going to replicate the test but it's taking a while to reindex zcashd and sync zebra, so I'll just approve it for now. I'll leave up to you if you'd like to merge this now, and how to handle the state issue. (I don't mind just merging it)

**Comment by teor2345** [comment]
> Took me a while to grok this (I always get confused by inputs/outputs) but I get it now, thanks!
> 
> I was going to replicate the test but it's taking a while to reindex zcashd and sync zebra, so I'll just approve it for now. I'll leave up to you if you'd like to merge this now, and how to handle the state issue. (I don't mind just merging it)

I have a bunch of release build fixes in PR #4369.
After they merge, I'll try to do a fix for the state rebuilds on top of those changes.
(We need to fix the state rebuild at some point, and we have another state version increment planned soon.)

**Comment by teor2345** [comment]
> CI is not rebuilding the cached state in this PR, even though the state version has changed.
> 
> So these tests failed:
> * fully-synced-rpc
> * lightwalletd-transactions

This should be fixed by PR #4385, which this PR is now based on.

**Comment by gustavovalverde** [comment]
I've updated this branch with https://github.com/ZcashFoundation/zebra/pull/4385 and `main` 

cc: @teor2345

**Comment by gustavovalverde** [comment]
https://github.com/ZcashFoundation/zebra/pull/4385 is working as expected now 🎉

File: .github/workflows/continous-integration-docker.yml [hunk 3]
```diff
@@ -66,23 +66,72 @@ env:
   GAR_BASE: us-docker.pkg.dev/zealous-zebra/zebra
 
 jobs:
-  changed-files:
+  get-available-disks:
```
Comment (dconnolly) [inline .github/workflows/continous-integration-docker.yml]: Update the .patch. one too?

Comment (teor2345) [reply .github/workflows/continous-integration-docker.yml]: This change needs to be made in PR #4385, @gustavovalverde can you do that?

File: .github/workflows/continous-integration-docker.yml [hunk 4]
```diff
@@ -245,7 +245,7 @@ jobs:
   regenerate-stateful-disks:
     needs: [ build, get-available-disks ]
     uses: ./.github/workflows/deploy-gcp-tests.yml
-    if: ${{ (!fromJSON(needs.get-available-disks.outputs.zebra_checkpoint_disk) && github.event_name == 'push') || github.event.inputs.regenerate-disks == 'true' }}
+    if: ${{ !fromJSON(needs.get-available-disks.outputs.zebra_checkpoint_disk) || github.event.inputs.regenerate-disks == 'true' }}
```
Comment (teor2345) [inline .github/workflows/continous-integration-docker.yml]: I don't know if these changes are due to a merge conflict, or they are actually in this PR.

Can we move them to PR #4385 @gustavovalverde ?

**Comment by gustavovalverde** [comment]
The full sync failed, but gracefully, from a CI perspective, as it happened what it was meant to happen (skipping dependent jobs).

As this was approved from a Rust/Zebra perspective, and the full sync issue is not related to this PR, I'm re-approving.

**Review by gustavovalverde — APPROVED** [review]
re-approving

**Pull request merged**